### PR TITLE
Generate out-of-time-bounds rejection log entries

### PR DESCRIPTION
**Pull request by miklos-da**
Summary of changes:
* Added field `outOfTimeBoundsLogEntry` to `CommitContext` to track rejection log entries and `preExecute` method to check if pre-execution logic needs to be enabled.
* Have all `Committer` implementations generate a rejection log entry in case pre-execution flag is set.
* Split `buildFinalResult` method in `TransactionCommitter` to make testing easier.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitterSpec.scala [hunk 1]
```diff
@@ -196,6 +196,35 @@ class TransactionCommitterSpec extends WordSpec with Matchers with MockitoSugar
     }
   }
 
+  "buildLogEntry" should {
+    "produce an out-of-time-bounds rejection log entry in case pre-execution is enabled" in {
+      val context = new FakeCommitContext(recordTime = None)
+
+      val actualSuccessLogEntry = instance.buildLogEntry(aTransactionEntrySummary, context)
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitterSpec.scala:203]: Why are the success log entry assertions present only in this test and not in others?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/ConfigCommitter.scala [hunk 2]
```diff
@@ -167,18 +167,25 @@ private[kvutils] class ConfigCommitter(
...
+        buildRejectionLogEntry(ctx, result.submission, identity, incrementMetric = false))
+    }
+    StepStop(successLogEntry)
   }
 
   private def buildRejectionLogEntry(
       ctx: CommitContext,
       submission: DamlConfigurationSubmission,
       addErrorDetails: DamlConfigurationRejectionEntry.Builder => DamlConfigurationRejectionEntry.Builder,
+      incrementMetric: Boolean = true,
```
[the remark this answers is not in the document]
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/ConfigCommitter.scala:184]: Well, that would make it even more confusing. The parameter `incrementMetric` has been introduced to allow reusing the code for building rejection log entries. I'll see if I can refactor this.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala [hunk 3]
```diff
@@ -144,9 +144,18 @@ private[kvutils] class PackageCommitter(
...
         .build
     )
-    val logEntry =
+    val successLogEntry =
       buildLogEntryWithOptionalRecordTime(ctx.getRecordTime, _.setPackageUploadEntry(uploadEntry))
-    StepStop(logEntry)
+    if (ctx.preExecute) {
+      ctx.outOfTimeBoundsLogEntry = Some(
+        buildRejectionLogEntry(
+          recordTime = None,
```
[the remark this answers is not in the document]
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala:152]: Added a separate method for setting out-of-time-bounds log entry.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PartyAllocationCommitter.scala [hunk 4]
```diff
@@ -118,18 +118,25 @@ private[kvutils] class PartyAllocationCommitter(
...
     )
 
-    val logEntry = buildLogEntryWithOptionalRecordTime(
+    val successLogEntry = buildLogEntryWithOptionalRecordTime(
       ctx.getRecordTime,
       _.setPartyAllocationEntry(partyAllocationEntry))
-    StepStop(logEntry)
+    if (ctx.preExecute) {
+      ctx.outOfTimeBoundsLogEntry = Some(
+        buildRejectionLogEntry(ctx, partyAllocationEntry, identity, incrementMetric = false))
```
[the remark this answers is not in the document]
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PartyAllocationCommitter.scala:126]: Added a separate method for setting out-of-time-bounds log entry.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitterSpec.scala (hunk 1 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitterSpec.scala:203]: Added test cases for checking just the success log entry independent from pre-execution (it's a separate concern).

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/ConfigCommitterSpec.scala [hunk 5]
```diff
@@ -94,6 +100,33 @@ class ConfigCommitterSpec extends WordSpec with Matchers {
...
+
+    "not set an out-of-time-bounds rejection log entry in case pre-execution is disabled" in {
+      val instance = createConfigCommitter(theRecordTime)
+      val context = new FakeCommitContext(recordTime = Some(aRecordTime))
+
+      instance.buildLogEntry(context, anEmptyResult)
+
+      context.preExecute shouldBe false
+      context.outOfTimeBoundsLogEntry shouldBe empty
+    }
```
[the remark this answers is not in the document]
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/ConfigCommitterSpec.scala:129]: Good point. However, I found the commonality so minimal that I'd rather keep the code easier to read than add another level of indirection.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala [hunk 6]
```diff
@@ -178,9 +184,9 @@ private[kvutils] class PackageCommitter(
       submissionId: String,
       participantId: String,
       addErrorDetails: DamlPackageUploadRejectionEntry.Builder => DamlPackageUploadRejectionEntry.Builder,
-      incrementMetric: Boolean = true,
+      incrementRejectionCount: Boolean = true,
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala]: I still feel like a less technical flag name would be more readable, like `dryRun` or `speculative` (so as not to tie this to `preexecution`) but there are less calls to this method now, so I'm OK with this.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala [hunk 7]
```diff
@@ -41,9 +45,15 @@ private[kvutils] trait CommitContext {
   var minimumRecordTime: Option[Instant] = None
   var maximumRecordTime: Option[Instant] = None
 
+  // Rejection log entry used for generating an out-of-time-bounds log entry in case of
+  // pre-execution.
+  var outOfTimeBoundsLogEntry: Option[DamlLogEntry] = None
```
[the remark this answers is not in the document]
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:50]: That's exactly the same comment I had on this before for you ;) mockito is happy to stub these as well so didn't make that change.

**Pull request merged**